folding-schemes/src/folding/traits.rs

```diff
@@ -0,0 +1,121 @@
+use ark_crypto_primitives::sponge::{
+    constraints::{AbsorbGadget, CryptographicSpongeVar},
+    poseidon::constraints::PoseidonSpongeVar,
+    Absorb,
+};
+use ark_ec::CurveGroup;
+use ark_ff::PrimeField;
+use ark_r1cs_std::{alloc::AllocVar, fields::fp::FpVar, ToConstraintFieldGadget};
+use ark_relations::r1cs::SynthesisError;
+
+use crate::{transcript::Transcript, Error};
+
+use super::circuits::CF1;
+
+pub trait CommittedInstanceExt<C: CurveGroup> {
+    /// The in-circuit representation of the committed instance.
+    type Var: AllocVar<Self, CF1<C>> + CommittedInstanceVarExt<C>;
+    /// `hash` implements the committed instance hash compatible with the
+    /// in-circuit implementation from `CommittedInstanceVarOps::hash`.
+    ///
+    /// Returns `H(i, z_0, z_i, U_i)`, where `i` can be `i` but also `i+1`, and
+    /// `U_i` is the committed instance `self`.
+    fn hash<T: Transcript<CF1<C>>>(
+        &self,
+        sponge: &T,
+        pp_hash: CF1<C>, // public params hash
+        i: CF1<C>,
+        z_0: &[CF1<C>],
+        z_i: &[CF1<C>],
+    ) -> CF1<C>
+    where
+        CF1<C>: Absorb,
+        Self: Sized + Absorb,
+    {
+        let mut sponge = sponge.clone();
+        sponge.absorb(&pp_hash);
+        sponge.absorb(&i);
+        sponge.absorb(&z_0);
+        sponge.absorb(&z_i);
+        sponge.absorb(&self);
+        sponge.squeeze_field_elements(1)[0]
+    }
+
+    /// Returns the commitments contained in the committed instance.
+    fn get_commitments(&self) -> Vec<C>;
+
+    /// Returns `true` if the committed instance is an incoming instance, and
+    /// `false` if it is a running instance.
+    fn is_incoming(&self) -> bool;
+
+    /// Checks if the committed instance is an incoming instance.
+    fn check_incoming(&self) -> Result<(), Error> {
```

Comment:
Question, I haven't found any usage of `is_incoming` nor `check_incoming` at the current branch, are they methods needed for later at #145 ?

Reply:
In fact it will be used in the [`refactor-arith-trait` branch](https://github.com/winderica/sonobe/tree/refactor-arith-trait), which is planned to be the second step to #145, but I haven't opened a PR for that.

Reply:
Cool, sounds good!